packages/textfield/src/mwc-textfield.ts

```diff
@@ -0,0 +1,293 @@
+  setValid(value: boolean): void;
+  setUseNativeValidation(value: boolean): void;
+  isDisabled(): boolean;
+  setDisabled(value: boolean): void;
+  setHelperTextContent(value: string): void;
+  setLeadingIconAriaLabel(value: string): void;
+  setLeadingIconContent(value: string): void;
+  setTrailingIconAriaLabel(value: string): void;
+  setTrailingIconContent(value: string): void;
+}
+
+export declare var TextFieldFoundation: {
+  prototype: TextFieldFoundation;
+  new (adapter: Adapter): TextFieldFoundation;
+}
+
+@customElement('mwc-textfield' as any)
+export class TextField extends BaseElement {
+  protected mdcFoundation!: TextFieldFoundation;
+
+  protected readonly mdcFoundationClass: typeof TextFieldFoundation = MDCTextFieldFoundation;
+
+  @query('.mdc-text-field')
+  protected mdcRoot!: HTMLElement;
+
+  @query(INPUT_SELECTOR)
+  protected inputSelector!: HTMLInputElement;
+
+  @query(LABEL_SELECTOR)
+  protected labelSelector!: HTMLElement;
+
+  @query(LINE_RIPPLE_SELECTOR)
+  protected lineRippleSelector!: HTMLElement;
+
+  @query(OUTLINE_SELECTOR)
+  protected outlineSelector!: HTMLElement;
+
+  @queryAll(ICON_SELECTOR)
+  protected iconSelector!: HTMLElement[];
+
+  @property({type: String})
+  value = ''
+
+  @property({type: String})
+  label = ''
+
+  @property({type: String})
+  icon = ''
+
+  @property({type: Boolean})
+  iconTrailing = false
+
+  @property({type: Boolean})
+  box = false
+
+  @property({type: Boolean})
+  outlined = false
+
+  @property({type: Boolean})
+  disabled = false
+
+  @property({type: Boolean})
+  fullWidth = false
+
+  @property({type: Boolean})
+  required = false
+
+  @property({type: String})
+  helperText = ''
+
+  @property({type: String})
+  placeHolder = ''
+
+  @property({type: String})
+  type = ''
+
+  @property()
+  lineRipple_
+
+  @property()
+  helperText_
+
+  @property()
+  leadingIcon_
+
+  @property()
+  trailingIcon_
+
+  @property()
+  label_
+
+  @property()
+  outline_
+
+  render() {
+    const classes = {
+      'mdc-text-field--with-leading-icon': this.icon && !this.iconTrailing,
+      'mdc-text-field--with-trailing-icon': this.icon && this.iconTrailing,
+      'mdc-text-field--box': this.box,
+      'mdc-text-field--outlined': !this.fullWidth && this.outlined,
+      'mdc-text-field--disabled': this.disabled,
+      'mdc-text-field--fullwidth': this.fullWidth,
+    };
+    return html`
+      ${this.renderStyle()}
+      <div class="mdc-text-field mdc-text-field--upgraded ${classMap(classes)}">
+        ${!this.fullWidth && this.icon ? html`<i class="material-icons mdc-text-field__icon" tabindex="0">${this.icon}</i>` : ''}
+        ${this._renderInput(this.value, this.required, this.type, this.placeHolder, this.label)}
+        ${!this.fullWidth ? html`<label class="mdc-floating-label ${this.value ? 'mdc-floating-label--float-above' : ''}" for="text-field">${this.label}</label>` : ''}
+        ${!this.fullWidth && this.outlined ? html`<div class="mdc-notched-outline">

```

Comment:
Nice work with the typescript version.
I think a small extension would be very handy when extending the textfield.
eg. with a password field with a button at the end of the field toggling content viability (type=input/password)
```suggestion
        <slot name="suffix" slot="suffix"></slot> 
        ${!this.fullWidth && this.outlined ? html`<div class="mdc-notched-outline">
```

Reply:
I looked into the changes, and I love the flexibility this provides, especially because I recently had to make use of textfields with a trailing icon and I would have loved this level of control. 

My solution was simply to make use of the "native" mwc-textfield way: the label property. I did have to manually add a click handler on the ```<i>``` tag provided, remove the fullWidth check, and be content with using material icons, but it was good enough for my needs. You can see it here:

![image](https://user-images.githubusercontent.com/11833744/49481855-8f60fb00-f7e1-11e8-8ad8-63712c003851.png)

markup before:
```html
${!this.fullWidth && this.icon ? html `<i class="material-icons mdc-text-field__icon" tabindex="0">${this.icon}</i>` : ''}
```

markup after:
```html
${this.icon ? html `<i @click="${this._onIconClick}" class="material-icons mdc-text-field__icon" tabindex="0">${this.icon}</i>` : ''}
```

As for integrating these changes, I think its best if we wait until a maintainer can go over the PR and your suggestion. Hopefully we can get an answer soon and have a more clear idea of what they want for this component.